### PR TITLE
[WIP]Bootstrap to update methods that can return nullptr

### PR DESCRIPTION
**Pull request by kumarak**

### CODE REVIEW

**Review by pgoodman — APPROVED** [review]

File: lib/AST/Bounds.cpp [hunk 1]
```diff
@@ -474,9 +474,14 @@ class DeclBoundsFinder : public clang::DeclVisitor<DeclBoundsFinder>,
...
 
       // NOTE(pag): This is a heuristic for detecting when things go "too far."
+      // NOTE(kumarak): We have gone too far and the token bound should end
+      //                if it reaches to a namespace. Return previous token in
+      //                that case
       if (tok_kind == clang::tok::kw_namespace) {
-        assert(kind == clang::tok::kw_namespace ||
-               tok[-1].Kind() == clang::tok::kw_using);
+        if (!(kind == clang::tok::kw_namespace ||
+            tok[-1].Kind() == clang::tok::kw_using)) {
```
Comment (pgoodman) [inline lib/AST/Bounds.cpp:482]: I think what's needed here is a helper method, like `PrevParsedToken`. It should return `nullptr` if there isn't a previous token. It should take in `tok`, and scan backwards, making sure to check if the token is actually a parsed token, i.e. `prev->IsParsed()`. Then, you can implement this check.

The key issue is that you could imagine a case involving some macro between `using` and `namespace`. If you notice other things that require looking at the previous token, then also update them to use this new method.

Having a `NextParsedToken` method would be good too.

**Review by pgoodman — CHANGES_REQUESTED** [review]

File: bin/BootstrapTypes/Globals.cpp [hunk 2]
```diff
@@ -1879,6 +1913,21 @@ std::map<std::pair<std::string, std::string>, std::string> kConditionalNullptr{
    "  if (self.isResultDependent()) {\n"
    "    return std::nullopt;\n"
    "  }\n"},
+  {{"PointerAttr", "DerefType"},
```
Comment (pgoodman) [inline bin/BootstrapTypes/Globals.cpp:1916]: We should have a rename for `Deref` to `Dereferenced`

Comment (pgoodman) [reply bin/BootstrapTypes/Globals.cpp:1916]: Then change this and below to `DereferencedType`

File: bin/BootstrapTypes/Globals.cpp [hunk 3]
```diff
@@ -1879,6 +1913,21 @@ std::map<std::pair<std::string, std::string>, std::string> kConditionalNullptr{
...
+   "  if (!self.getDerefTypeLoc()) {\n"
+   "    return std::nullopt;\n"
+   "  }\n"},
+  {{"OwnerAttr", "DerefType"},
```
Comment (pgoodman) [inline bin/BootstrapTypes/Globals.cpp:1920]: I thought we don't try to get `Type`s from `TypeLoc`s anymore?

File: include/pasta/AST/Stmt.h [hunk 4]
```diff
@@ -1136,13 +1136,13 @@ class CXXForRangeStmt : public Stmt {
   PASTA_DECLARE_BASE_OPERATORS(Stmt, CXXForRangeStmt)
   std::vector<::pasta::Stmt> Children(void) const;
   ::pasta::Token BeginToken(void) const;
-  ::pasta::DeclStmt BeginStatement(void) const;
+  std::optional<::pasta::DeclStmt> BeginStatement(void) const;
```
Comment (pgoodman) [inline include/pasta/AST/Stmt.h:1139]: What are `CXXCatchStmt::BeginStatement` and `EndStatement`? In Clang I imagine we can use these for iteration, but probably not here?

**Review by pgoodman — CHANGES_REQUESTED** [review]

File: bin/BootstrapTypes/Globals.cpp (hunk 3 above)
Comment (kumarak) [reply bin/BootstrapTypes/Globals.cpp:1920]: We do get it in some cases and generate the pasta type.

File: include/pasta/AST/Stmt.h (hunk 4 above)
Comment (kumarak) [reply include/pasta/AST/Stmt.h:1139]: In clang we can use it to get `Begin` and `End` stmt during iterations and here PASTA apis provide the wrapped begin and end stmt. It gets used during serialization of `CXXForRangeStmt`.

**Pull request merged**